File: drestaurant-libs/drestaurant-order/src/test/java/com/drestaurant/order/domain/OrderSagaTest.java
package com.drestaurant.order.domain;

import java.util.ArrayList;
import java.util.List;

import org.axonframework.test.saga.FixtureConfiguration;
import org.axonframework.test.saga.SagaTestFixture;
import org.junit.Before;
import org.junit.Test;

import com.drestaurant.common.domain.model.AuditEntry;
import com.drestaurant.common.domain.model.Money;
import com.drestaurant.courier.domain.api.CourierOrderCreatedEvent;
import com.drestaurant.courier.domain.api.CourierOrderDeliveredEvent;
import com.drestaurant.courier.domain.api.CreateCourierOrderCommand;
import com.drestaurant.customer.domain.api.CreateCustomerOrderCommand;
import com.drestaurant.customer.domain.api.CustomerOrderCreatedEvent;
import com.drestaurant.customer.domain.api.CustomerOrderRejectedEvent;
import com.drestaurant.customer.domain.api.MarkCustomerOrderAsDeliveredCommand;
import com.drestaurant.order.domain.api.OrderCreationInitiatedEvent;
import com.drestaurant.order.domain.api.OrderDeliveredEvent;
import com.drestaurant.order.domain.api.OrderPreparedEvent;
import com.drestaurant.order.domain.api.OrderVerifiedByCustomerEvent;
import com.drestaurant.order.domain.model.OrderDetails;
import com.drestaurant.order.domain.model.OrderInfo;
import com.drestaurant.order.domain.model.OrderLineItem;
import com.drestaurant.restaurant.domain.api.CreateRestaurantOrderCommand;
import com.drestaurant.restaurant.domain.api.RestaurantOrderCreatedEvent;
import com.drestaurant.restaurant.domain.api.RestaurantOrderPreparedEvent;
import com.drestaurant.restaurant.domain.api.RestaurantOrderRejectedEvent;
import com.drestaurant.restaurant.domain.model.RestaurantOrderDetails;
import com.drestaurant.restaurant.domain.model.RestaurantOrderLineItem;

public class OrderSagaTest {

	private FixtureConfiguration testFixture;
	private AuditEntry auditEntry;
	private static final String WHO = "johndoe";
	private String orderId;
	private String customerId;
	private String restaurantId;
	
	private List<OrderLineItem> lineItems;
	private OrderLineItem lineItem1;
	private OrderLineItem lineItem2;
	private OrderInfo orderInfo;
	private OrderDetails orderDetails;
	private List<RestaurantOrderLineItem> restaurantLineItems;
	private RestaurantOrderDetails restaurantOrderDetails;

	@Before
	public void setUp() throws Exception {
		testFixture = new SagaTestFixture<>(OrderSaga.class);
		orderId = "orderId";
		customerId = "customerId";
		restaurantId = "restaurantId";
		
		auditEntry = new AuditEntry(WHO);		
		lineItems = new ArrayList<OrderLineItem>();
		lineItem1 = new OrderLineItem("menuItemId1","name1", new Money(11), 2);
		lineItem2 = new OrderLineItem("menuItemId2","name2", new Money(22), 3);
		lineItems.add(lineItem1);
		lineItems.add(lineItem2);
		orderInfo = new OrderInfo(customerId, restaurantId, lineItems);
		orderDetails = new OrderDetails(orderInfo, lineItems.stream().map(OrderLineItem::getTotal).reduce(Money.ZERO, Money::add));
		
		restaurantLineItems = new ArrayList<>();
		for(OrderLineItem oli : this.orderDetails.getLineItems()) {
			RestaurantOrderLineItem roli = new RestaurantOrderLineItem(oli.getQuantity(), oli.getMenuItemId(), oli.getName());
			restaurantLineItems.add(roli);
		}
	    restaurantOrderDetails = new RestaurantOrderDetails(restaurantLineItems);
		
	}
	
	@Test
	public void orderCreationInitiatedTest() throws Exception {

		testFixture.givenNoPriorActivity()
		           .whenAggregate(orderId)
				   .publishes(
						   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
						   )
				   .expectActiveSagas(1)
				   .expectDispatchedCommands(new CreateCustomerOrderCommand("customerOrder_" + orderId, orderDetails.getOrderTotal(), customerId, auditEntry));
	}
	
	@Test
	public void customerOrderCreatedTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
		        		   )
		           .whenPublishingA(new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry))
		           .expectActiveSagas(1)
		           .expectDispatchedCommands(new MarkOrderAsVerifiedByCustomerCommand(orderId, customerId, auditEntry));
	}
	
	@Test
	public void orderVerifiedByCustomerTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
		        		   )
		           .whenPublishingA(new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry))
		           .expectActiveSagas(1)
		           .expectDispatchedCommands(new CreateRestaurantOrderCommand("restaurantOrder_" + orderId, restaurantOrderDetails, restaurantId, auditEntry));
	}
	
	@Test
	public void restaurantOrderCreatedEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
		        		   new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry),
		        		   new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry)
		        		   )
		           .whenPublishingA(new RestaurantOrderCreatedEvent("restaurantOrder_" + orderId, auditEntry))
		           .expectActiveSagas(1)
		           .expectDispatchedCommands(new MarkOrderAsVerifiedByRestaurantCommand(orderId, restaurantId, auditEntry));
	}
	
	@Test
	public void restaurantOrderPreparedEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
		        		   new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry),
		        		   new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
		        		   new RestaurantOrderCreatedEvent("restaurantOrder_"+ orderId, auditEntry)
		        		   )
		           .whenPublishingA(new RestaurantOrderPreparedEvent("restaurantOrder_" + orderId, auditEntry))
		           .expectActiveSagas(1)
		           .expectDispatchedCommands(new MarkOrderAsPreparedCommand(orderId, auditEntry));
	}
	@Test
	public void orderPreparedEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
		        		   new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry),
		        		   new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
		        		   new RestaurantOrderCreatedEvent("restaurantOrder_" + orderId, auditEntry),
		        		   new RestaurantOrderPreparedEvent("restaurantOrder_" + orderId, auditEntry)
		        		   )
		           .whenPublishingA(new OrderPreparedEvent(orderId, auditEntry))
		           .expectActiveSagas(1)
		           .expectDispatchedCommands(new CreateCourierOrderCommand("courierOrder_" + orderId, auditEntry));
	}
	@Test
	public void courierOrderCreatedEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
		        		   new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry),
		        		   new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
		        		   new RestaurantOrderCreatedEvent("restaurantOrder_" + orderId, auditEntry),
		        		   new RestaurantOrderPreparedEvent("restaurantOrder_" + orderId, auditEntry),
		        		   new OrderPreparedEvent(orderId, auditEntry)
		        		   )
		           .whenPublishingA(new CourierOrderCreatedEvent("courierOrder_" + orderId, auditEntry))
		           .expectActiveSagas(1)
		           .expectDispatchedCommands(new MarkOrderAsReadyForDeliveryCommand(orderId, auditEntry));
	}
	
	@Test
	public void courierOrderDeliveredEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
		        		   new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry),
		        		   new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
		        		   new RestaurantOrderCreatedEvent("restaurantOrder_" + orderId, auditEntry),
		        		   new RestaurantOrderPreparedEvent("restaurantOrder_" + orderId, auditEntry),
		        		   new OrderPreparedEvent(orderId, auditEntry),
		        		   new CourierOrderCreatedEvent("courierOrder_" + orderId, auditEntry)
		        		   )
		           .whenPublishingA(new CourierOrderDeliveredEvent("courierOrder_" + orderId, auditEntry))
		           .expectActiveSagas(1)
		           .expectDispatchedCommands(new MarkOrderAsDeliveredCommand(orderId, auditEntry));
	}
	
	@Test
	public void customerOrderDeliveredEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
		        		   new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry),
		        		   new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
		        		   new RestaurantOrderCreatedEvent("restaurantOrder_" + orderId, auditEntry),
		        		   new RestaurantOrderPreparedEvent("restaurantOrder_" + orderId, auditEntry),
		        		   new OrderPreparedEvent(orderId, auditEntry),
		        		   new CourierOrderCreatedEvent("courierOrder_" + orderId, auditEntry),
		        		   new CourierOrderDeliveredEvent("courierOrder_" + orderId, auditEntry)
		        		   )
		           .whenPublishingA(new OrderDeliveredEvent(orderId, auditEntry))
		           .expectActiveSagas(0)
		           .expectDispatchedCommands(new MarkCustomerOrderAsDeliveredCommand("customerOrder_" + orderId, auditEntry));
	}
	
	@Test
	public void customerOrderRejectedEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
		        		   )
		           .whenPublishingA(new CustomerOrderRejectedEvent("customerOrder_" + orderId, auditEntry))
		           .expectActiveSagas(0)
		           .expectDispatchedCommands(new MarkOrderAsRejectedCommand(orderId, auditEntry));
	}
	
	@Test
	public void restaurantOrderRejectedEventTest() throws Exception {

		testFixture.givenAggregate(orderId)
		           .published(
		        		   new OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
		        		   new CustomerOrderCreatedEvent("customerOrder_" + orderId, auditEntry),
		        		   new OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry)
		        		   )
		           .whenPublishingA(new RestaurantOrderRejectedEvent("restaurantOrder_" + orderId, auditEntry))
		           .expectActiveSagas(0)
		           .expectDispatchedCommands(new MarkOrderAsRejectedCommand(orderId, auditEntry));
	}
}

File: drestaurant-libs/drestaurant-order/src/test/kotlin/com/drestaurant/order/domain/OrderSagaTest.kt
package com.drestaurant.order.domain

import com.drestaurant.common.domain.api.model.AuditEntry
import com.drestaurant.common.domain.api.model.Money
import com.drestaurant.courier.domain.api.CourierOrderCreatedEvent
import com.drestaurant.courier.domain.api.CourierOrderDeliveredEvent
import com.drestaurant.courier.domain.api.CreateCourierOrderCommand
import com.drestaurant.courier.domain.api.model.CourierOrderId
import com.drestaurant.customer.domain.api.CreateCustomerOrderCommand
import com.drestaurant.customer.domain.api.CustomerOrderCreatedEvent
import com.drestaurant.customer.domain.api.CustomerOrderRejectedEvent
import com.drestaurant.customer.domain.api.model.CustomerId
import com.drestaurant.customer.domain.api.model.CustomerOrderId
import com.drestaurant.order.domain.api.OrderCreationInitiatedEvent
import com.drestaurant.order.domain.api.OrderPreparedEvent
import com.drestaurant.order.domain.api.OrderVerifiedByCustomerEvent
import com.drestaurant.order.domain.api.model.OrderDetails
import com.drestaurant.order.domain.api.model.OrderId
import com.drestaurant.order.domain.api.model.OrderInfo
import com.drestaurant.order.domain.api.model.OrderLineItem
import com.drestaurant.restaurant.domain.api.CreateRestaurantOrderCommand
import com.drestaurant.restaurant.domain.api.RestaurantOrderCreatedEvent
import com.drestaurant.restaurant.domain.api.RestaurantOrderPreparedEvent
import com.drestaurant.restaurant.domain.api.RestaurantOrderRejectedEvent
import com.drestaurant.restaurant.domain.api.model.RestaurantId
import com.drestaurant.restaurant.domain.api.model.RestaurantOrderDetails
import com.drestaurant.restaurant.domain.api.model.RestaurantOrderId
import com.drestaurant.restaurant.domain.api.model.RestaurantOrderLineItem
import org.axonframework.test.saga.FixtureConfiguration
import org.axonframework.test.saga.SagaTestFixture
import org.junit.Before
import org.junit.Test
import java.math.BigDecimal
import java.util.*

class OrderSagaTest {

    private lateinit var testFixture: FixtureConfiguration

    private val who = "johndoe"
    private var auditEntry: AuditEntry = AuditEntry(who, Calendar.getInstance().time)

    private val lineItems: MutableList<OrderLineItem> = ArrayList()
    private val lineItem1: OrderLineItem = OrderLineItem("menuItemId1", "name1", Money(BigDecimal.valueOf(11)), 2)
    private val lineItem2: OrderLineItem = OrderLineItem("menuItemId2", "name2", Money(BigDecimal.valueOf(22)), 3)
    private var orderInfo: OrderInfo = OrderInfo("customerId", "restaurantId", lineItems)
    private val orderDetails: OrderDetails = OrderDetails(orderInfo, lineItems.stream().map(OrderLineItem::total).reduce(Money(BigDecimal.ZERO), Money::add))
    private val orderId: OrderId = OrderId("orderId")
    private var customerId: CustomerId = CustomerId("customerId")
    private var restaurantId: RestaurantId = RestaurantId("restaurantId")

    private var restaurantLineItems: MutableList<RestaurantOrderLineItem> = ArrayList()
    private var restaurantOrderDetails: RestaurantOrderDetails = RestaurantOrderDetails(restaurantLineItems)

    @Before
    fun setUp() {

        testFixture = SagaTestFixture(OrderSaga::class.java)
        lineItems.add(lineItem1)
        lineItems.add(lineItem2)

        for (oli in this.orderDetails.lineItems) {
            val roli = RestaurantOrderLineItem(oli.quantity, oli.menuItemId, oli.name)
            restaurantLineItems.add(roli)
        }
    }

    @Test
    fun orderCreationInitiatedTest() {

        testFixture.givenNoPriorActivity()
                .whenAggregate(orderId.toString())
                .publishes(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
                )
                .expectActiveSagas(1)
                .expectDispatchedCommands(CreateCustomerOrderCommand(customerId, CustomerOrderId("customerOrder_$orderId"), orderDetails.orderTotal, auditEntry))
    }

    @Test
    fun customerOrderCreatedTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
                )
                .whenPublishingA(CustomerOrderCreatedEvent(orderDetails.orderTotal, customerId, CustomerOrderId("customerOrder_$orderId"), auditEntry))
                .expectActiveSagas(1)
                .expectDispatchedCommands(MarkOrderAsVerifiedByCustomerInternalCommand(orderId, customerId, auditEntry))
    }

    @Test
    fun orderVerifiedByCustomerTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
                )
                .whenPublishingA(OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry))
                .expectActiveSagas(1)
                .expectDispatchedCommands(CreateRestaurantOrderCommand(restaurantId, restaurantOrderDetails, RestaurantOrderId("restaurantOrder_$orderId"), auditEntry))
    }

    @Test
    fun restaurantOrderCreatedEventTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
                        CreateCustomerOrderCommand(customerId, CustomerOrderId("customerOrder_$orderId"), orderDetails.orderTotal, auditEntry),
                        OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry)
                )
                .whenPublishingA(RestaurantOrderCreatedEvent(restaurantLineItems, RestaurantOrderId("restaurantOrder_$orderId"), restaurantId, auditEntry))
                .expectActiveSagas(1)
                .expectDispatchedCommands(MarkOrderAsVerifiedByRestaurantInternalCommand(orderId, restaurantId, auditEntry))
    }

    @Test
    fun restaurantOrderPreparedEventTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
                        CreateCustomerOrderCommand(customerId, CustomerOrderId("customerOrder_$orderId"), orderDetails.orderTotal, auditEntry),
                        OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
                        RestaurantOrderCreatedEvent(restaurantLineItems, RestaurantOrderId("restaurantOrder_$orderId"), restaurantId, auditEntry)
                )
                .whenPublishingA(RestaurantOrderPreparedEvent(RestaurantOrderId("restaurantOrder_$orderId"), auditEntry))
                .expectActiveSagas(1)
                .expectDispatchedCommands(MarkOrderAsPreparedInternalCommand(orderId, auditEntry))
    }

    @Test
    fun orderPreparedEventTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
                        CreateCustomerOrderCommand(customerId, CustomerOrderId("customerOrder_$orderId"), orderDetails.orderTotal, auditEntry),
                        OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
                        RestaurantOrderCreatedEvent(restaurantLineItems, RestaurantOrderId("restaurantOrder_$orderId"), restaurantId, auditEntry),
                        RestaurantOrderPreparedEvent(RestaurantOrderId("restaurantOrder_$orderId"), auditEntry)
                )
                .whenPublishingA(OrderPreparedEvent(orderId, auditEntry))
                .expectActiveSagas(1)
                .expectDispatchedCommands(CreateCourierOrderCommand(CourierOrderId("courierOrder_$orderId"), auditEntry))
    }

    @Test
    fun courierOrderCreatedEventTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
                        CreateCustomerOrderCommand(customerId, CustomerOrderId("customerOrder_$orderId"), orderDetails.orderTotal, auditEntry),
                        OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
                        RestaurantOrderCreatedEvent(restaurantLineItems, RestaurantOrderId("restaurantOrder_$orderId"), restaurantId, auditEntry),
                        RestaurantOrderPreparedEvent(RestaurantOrderId("restaurantOrder_$orderId"), auditEntry),
                        OrderPreparedEvent(orderId, auditEntry)
                )
                .whenPublishingA(CourierOrderCreatedEvent(CourierOrderId("courierOrder_$orderId"), auditEntry))
                .expectActiveSagas(1)
                .expectDispatchedCommands(MarkOrderAsReadyForDeliveryInternalCommand(orderId, auditEntry))
    }

    @Test
    fun courierOrderDeliveredEventTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
                        CreateCustomerOrderCommand(customerId, CustomerOrderId("customerOrder_$orderId"), orderDetails.orderTotal, auditEntry),
                        OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry),
                        RestaurantOrderCreatedEvent(restaurantLineItems, RestaurantOrderId("restaurantOrder_$orderId"), restaurantId, auditEntry),
                        RestaurantOrderPreparedEvent(RestaurantOrderId("restaurantOrder_$orderId"), auditEntry),
                        OrderPreparedEvent(orderId, auditEntry),
                        CourierOrderCreatedEvent(CourierOrderId("courierOrder_$orderId"), auditEntry)
                )
                .whenPublishingA(CourierOrderDeliveredEvent(CourierOrderId("courierOrder_$orderId"), auditEntry))
                .expectActiveSagas(0)
                .expectDispatchedCommands(MarkOrderAsDeliveredInternalCommand(orderId, auditEntry))
    }

    @Test
    fun customerOrderRejectedEventTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry)
                )
                .whenPublishingA(CustomerOrderRejectedEvent(CustomerOrderId("customerOrder_$orderId"), auditEntry))
                .expectActiveSagas(0)
                .expectDispatchedCommands(MarkOrderAsRejectedInternalCommand(orderId, auditEntry))
    }

    @Test
    fun restaurantOrderRejectedEventTest() {

        testFixture.givenAggregate(orderId.toString())
                .published(
                        OrderCreationInitiatedEvent(orderDetails, orderId, auditEntry),
                        CreateCustomerOrderCommand(customerId, CustomerOrderId("customerOrder_$orderId"), orderDetails.orderTotal, auditEntry),
                        OrderVerifiedByCustomerEvent(orderId, customerId, auditEntry)
                )
                .whenPublishingA(RestaurantOrderRejectedEvent(RestaurantOrderId("restaurantOrder_$orderId"), auditEntry))
                .expectActiveSagas(0)
                .expectDispatchedCommands(MarkOrderAsRejectedInternalCommand(orderId, auditEntry))
    }
}
